### PR TITLE
XOL-4081 Smartwaiver plugin app

### PR DESCRIPTION
**Pull request by jeetsang**

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import morgan from 'morgan'
+
+const app = express()
+
+app.use(express.json())
+app.use(express.urlencoded({extended: false}))
+app.use(cookieParser())
+app.use(bodyParser.json())
+app.use(express.static(path.join(__dirname, 'public')))
+app.use(morgan('combined', {'stream': logger.stream}))
```
Comment (anushr) [inline app.js]: What about logentries?

File: controllers/appStoreWebhookController.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
+import createError from 'http-errors'
+import registrationService from './../services/registrationService'
+
+const AppStoreWebhookController = {
+  processWebhook: async (req, res) => {
+    const {eventName, data} = req.body.payload
+    switch (eventName) {
+      case 'plugin.install':
```
Comment (anushr) [inline controllers/appStoreWebhookController.js]: This event is not fired in elrond. Listen to `installation.create`

File: controllers/appStoreWebhookController.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+        await registrationService.register(data)
+        res.status(200).send()
+        break
+      case 'plugin.uninstall' :
```
Comment (anushr) [inline controllers/appStoreWebhookController.js]: `installation.delete`

File: logger.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import Logger from 'le_node' // Do not remove; This adds a transport to winston.
+import config from 'config'
+import winston from 'winston'
+
+const logger = new winston.Logger({
+  level: 'info',
+  json: true
+})
+if (process.env.NODE_ENV !== 'test') {
```
Comment (anushr) [inline logger.js]: Why is this check here? The config alone should dictate what happens. No environment specific hardcoding.

File: package.json [hunk 5]
```diff
@@ -0,0 +1,45 @@
+{
+  "name": "smartwaiver-app",
+  "version": "0.0.0",
+  "private": true,
+  "scripts": {
+    "start": "nodemon ./bin/www --exec babel-node --presets es2015,stage-2",
+    "devStart": "nodemon ./bin/www --exec babel-node --presets es2015,stage-2",
```
Comment (anushr) [inline package.json]: No babel. Use esm for `import` support.

https://github.com/standard-things/esm

File: package.json [hunk 6]
```diff
@@ -0,0 +1,45 @@
+{
+  "name": "smartwaiver-app",
+  "version": "0.0.0",
+  "private": true,
+  "scripts": {
+    "start": "nodemon ./bin/www --exec babel-node --presets es2015,stage-2",
```
Comment (anushr) [inline package.json]: No nodemon for `start`.

File: bin/www [hunk 7]
```diff
@@ -0,0 +1,91 @@
+#!/usr/bin/env node
```
Comment (anushr) [inline bin/www]: Remove this file.

File: controllers/appStoreWebhookController.js [hunk 8]
```diff
@@ -0,0 +1,24 @@
+import createError from 'http-errors'
+import registrationService from './../services/registrationService'
+
+const AppStoreWebhookController = {
```
Comment (anushr) [inline controllers/appStoreWebhookController.js]: Filename for classes should be TitleCase, not camelCase. Do this for the entire PR.

File: controllers/appStoreWebhookController.js [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+        break
+      default:
+        throw createError(400, {}, {
+          'payload.eventName': 'invalid',
+          message: 'Your request contains an invalid value'
+        })
+    }
+  }
+}
+module.exports = AppStoreWebhookController
```
Comment (anushr) [inline controllers/appStoreWebhookController.js]: Add `;` to all `module.export` lines for entire PR.

File: logger.js (hunk 4 above)
Comment (srishaharidas) [reply logger.js]: removed

File: controllers/appStoreWebhookController.js (hunk 2 above)
Comment (srishaharidas) [reply controllers/appStoreWebhookController.js]: this was in the todo, have changed this now

File: controllers/appStoreWebhookController.js (hunk 3 above)
Comment (srishaharidas) [reply controllers/appStoreWebhookController.js]: changed

File: controllers/appStoreWebhookController.js (hunk 8 above)
Comment (srishaharidas) [reply controllers/appStoreWebhookController.js]: done

File: controllers/appStoreWebhookController.js (hunk 9 above)
Comment (srishaharidas) [reply controllers/appStoreWebhookController.js]: done

File: repositories/userConfigRepository.js [hunk 10]
```diff
@@ -0,0 +1,62 @@
+import createError from 'http-errors'
+import _ from 'lodash'
+import NodeCache from 'node-cache'
+import {SW_API_KEY_NAME, ENCODING_HEX} from '../constants/constants'
+import logger from '../logger'
+import encryptionService from '../services/encryptionService'
+import xola from '../services/xolaService'
+const cache = new NodeCache()
+
+const repository = {
```
[the remark this answers is not in the document]
Comment (srishaharidas) [reply repositories/userConfigRepository.js]: done

File: app.js [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+  // set locals, only providing error in development
+  res.locals.message = err.message
+  res.locals.error = req.app.get('env') === 'development' ? err : {}
+
+  // render the error page
+  res.status(err.status || 500)
+  res.send(err)
+})
+
+module.exports = app
```
[the remark this answers is not in the document]
Comment (srishaharidas) [reply app.js]: changed

File: authentication/authenticator.js [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+const authenticator = {
+  authenticate: (req, res, next) => {
+    if (req.headers['authentication-key'] !== config.get('xola.apiKey')) {
+      throw createError(401)
+    }
+    next()
+  }
+}
+
+module.exports = authenticator
```
[the remark this answers is not in the document]
Comment (srishaharidas) [reply authentication/authenticator.js]: changed

File: constants/constants.js [hunk 13]
```diff
@@ -0,0 +1,7 @@
+const constants = {
+  SW_API_KEY_NAME: 'smart_waiver_api_key',
+  ENCODING_HEX: 'hex',
+  ENCODING_UTF: 'utf8',
+  ENCRYPTION_ALGORITHM_AES_256_CBC: 'aes-256-cbc'
+}
+module.exports = constants
```
[the remark this answers is not in the document]
Comment (srishaharidas) [reply constants/constants.js]: changed

File: controllers/appStoreWebhookController.js (hunk 8 above)
Comment (srishaharidas) [reply controllers/appStoreWebhookController.js]: changed to TitleCase

File: index.js [hunk 14]
```diff
@@ -0,0 +1,10 @@
+// Enable ES module loader.
+require = require('esm')(module);
+const config = require('config');
+const PORT = config.get("app.port");
+
```
Comment (nemanjakrstic) [inline index.js]: Yesterday we encountered a problem with using `esm` and `config` packages like this.

Here the problem. If you first use `require` to load `config` package, and then in some other file you use `import` to load same package, package will resolve in an empty object.

Solution is to make new entry point for your app, for example `index.esm.js`, containing only the following:
```
require = require('esm')(module);
require('./index');
```
And from there, index.js can contain ES module imports:
```
import config from 'config';
...rest of the code...
```

Then, just point all scripts to `index.esm.js` in your package.json.

Ok?

File: services/EncryptionService.js [hunk 15]
```diff
@@ -0,0 +1,28 @@
+import crypto from 'crypto';
+import { ENCRYPTION_ALGORITHM_AES_256_CBC, ENCODING_HEX, ENCODING_UTF } from '../constants/constants';
+const config = require('config');
```
Comment (nemanjakrstic) [inline services/EncryptionService.js]: Use ES imports, they will work now.

File: services/RegistrationService.js [hunk 16]
```diff
@@ -0,0 +1,59 @@
+import _ from "lodash";
+import createError from "http-errors";
+import {SW_API_KEY_NAME} from "../constants/constants";
+import Logger from "../Logger";
+import userConfigRepository from "../repositories/UserConfigRepository";
+import smartWaiverService from "./SmartWaiverService";
+import xolaService from "./XolaService";
+const config = require('config');
```
Comment (nemanjakrstic) [inline services/RegistrationService.js]: Use ES imports. Do the same for all files.

File: app.js [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+import bodyParser from 'body-parser';
+import router from './router';
+import morgan from 'morgan';
+
+const app = express();
+
+app.use(express.json());
+app.use(express.urlencoded({ extended: false }));
+app.use(cookieParser());
+app.use(bodyParser.json());
```
[the remark this answers is not in the document]
Comment (srishaharidas) [reply app.js]: you're right, i've removed ` bodyParser.json() ` have kept  express.json()

File: index.js (hunk 14 above)
Comment (srishaharidas) [reply index.js]: right, got it. was struggling this. thanks, this fixed it!

File: services/EncryptionService.js (hunk 15 above)
Comment (srishaharidas) [reply services/EncryptionService.js]: done

File: services/RegistrationService.js (hunk 16 above)
Comment (srishaharidas) [reply services/RegistrationService.js]: done

**Pull request closed without merging**